File: src/main/java/leetcode/tree/PathSumThree.java
package leetcode.tree;

import java.util.*;

public class PathSumThree {
  int target;
  int count = 0;
  Map<Integer, Integer> counting = new HashMap<>();

  public int pathSum(TreeNode root, int sum) {
    this.target = sum;
    traverse(root, 0);
    return count;
  }

  public void traverse(TreeNode root, int sum) {
    if (root == null)
      return;

    sum += root.val;

    if (sum == this.target)
      count++;

    if (counting.containsKey(sum - target))
      count += counting.get(sum - target);

    counting.put(sum, counting.getOrDefault(sum, 0) + 1);

    traverse(root.left, sum);
    traverse(root.right, sum);

    counting.put(sum, counting.get(sum) - 1);
  }
}

File: src/main/python/leetcode/tree/path_sum_three.py
class Solution:
    def pathSum(self, root: TreeNode, sum: int) -> int:
        self.target = sum
        self.count = 0
        self.h = {}
        self.traverse(root, 0)
        return self.count

    def traverse(self, root: TreeNode, sum: int):
        if root is None:
            return

        sum += root.val
        if sum == self.target:
            self.count = self.count + 1

        if sum not in self.h:
            self.h[sum] = 0
    
        if (sum - self.target) in self.h:
            self.count += self.h[sum - self.target]

        self.h[sum] = self.h[sum] + 1
        self.traverse(root.left, sum)
        self.traverse(root.right, sum)
        self.h[sum] = self.h[sum] - 1
